syncthing: Add discovery/relay servers, improve build

andrew-d opened the pull request:
###### Motivation for this change
I wanted to be able to run the Syncthing discovery / relay servers on my own infrastructure.  In the process, I also fixed the silly `$GOPATH` check that the Syncthing build does (see [here][bf]), and included man pages in the package.

[bf]: https://github.com/syncthing/syncthing/blob/v0.14.44/build.go#L208-L220

cc @pshendry @jokogr @peterhoeg (existing maintainers of this package)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] ~Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))~
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

andrew-d committed:
syncthing: fix GOPATH checks

andrew-d committed:
syncthing: Install manpages

andrew-d committed:
syncthing: Also build and install discovery and relay servers

peterhoeg commented:
I'm not sure if the use of multiple outputs to do this is the right way. The "proper" way IMHO would be to create separate derivations (and an additional "cli" derivation for the ```stcli``` tool).

andrew-d commented:
Sure, happy to switch to separate derivations if you'd prefer - mind pointing me to an example of what this would look like (e.g. a package that does this)?

andrew-d commented:
(I was also considering just not bothering splitting things up and building everything in the main package, since the build is only a couple seconds slower here - thoughts on that approach?)

peterhoeg commented:
> mind pointing me to an example of what this would look like (e.g. a package that does this)?

Here one shared source is used by multiple derivations: https://github.com/NixOS/nixpkgs/pull/34611

> building everything in the main package

From my point of view - "it depends".  The go binaries are pretty big so the nice thing to do is to split it up and people can pull in just what they need.

andrew-d committed:
Review feedback: split into different derivations

andrew-d commented:
@peterhoeg - Apologies for the wait on this one, but I just pushed a commit that splits up things into a common core, and different packages for `syncthing-relay` and `syncthing-discovery`.  Thoughts on this approach?

peterhoeg commented:
Nothing to apologize for - we all get busy with other things.

It makes sense to split them up if they are likely to be used independently.

As for the approach using ```common.nix``` - there is technically nothing wrong with it, it just isn't how most of the other packages are done. Have a look at https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/cudatoolkit/default.nix as an example of using a function.

andrew-d committed:
Review feedback: keep common config in a single file

andrew-d commented:
Ah, that's helpful to see - thanks!  Just pushed another commit that uses a `common` function in a single file.  The [diff without whitespace](https://github.com/NixOS/nixpkgs/pull/34831/files?w=1) might be helpful.

peterhoeg commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -1,57 +1,99 @@
-  };
+    stdenv.mkDerivation rec {
+      version = "0.14.44";
+      name = "${args.name}-${version}";
```
You can just use ```name``` instead of ```args.name```.

andrew-d replied:
I wasn't able to do this, since I got an infinite recursion error (presumably with the value of the `name` attribute itself); instead, I renamed the parameter.

peterhoeg commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -1,57 +1,99 @@
+        sha256 = "1gdkx6lbzmdz2hqc9slbq41rwgkxmdisnj0iywx4mppmc2b4v6wh";
+      };
+
+      buildInputs = [ go removeReferencesTo ];
```
```removeReferencesTo``` should be in ```nativeBuildInputs```.

peterhoeg commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -1,57 +1,99 @@
+        env GOPATH= go run build.go -no-upgrade -version v${version} build ${target}
+      '';
+
+      inherit (args) installPhase;
```
You're passing in ```installPhase``` so you can do ```inherit installPhase```.

peterhoeg commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -1,57 +1,99 @@
+        homepage = https://www.syncthing.net/;
+        description = "Open Source Continuous File Synchronization";
+        license = licenses.mpl20;
+        maintainers = with maintainers; [ pshendry joko peterhoeg ];
```
You don't want to add yourself?

andrew-d replied:
I wasn't sure what the etiquette here is - if you're okay with it, I've added myself.

peterhoeg replied:
Absolutely! We need all the maintainers we can get.

peterhoeg commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -1,57 +1,99 @@
+      mkdir -p $out/lib/systemd/{system,user}
 
-  installPhase = ''
-    mkdir -p $out/lib/systemd/{system,user}
```
This should move into the linux specific section below.

peterhoeg commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -1,57 +1,99 @@
-    find $out/bin -type f -exec remove-references-to -t ${go} '{}' '+'
-  '';
+    installPhase = ''
+      mkdir -p $out/lib/systemd/system $out/bin
```
This should also move down.

peterhoeg commented on pkgs/top-level/all-packages.nix:
```diff
@@ -17327,7 +17327,10 @@ with pkgs;
+  inherit (callPackages ../applications/networking/syncthing { })
+    syncthing
+    syncthing-discovery
+    syncthing-relay;
```
Please also create ```syncthing-cli``` with ```stcli``` inside.

andrew-d replied:
Done!  It needed a small patch since it wasn't a build target by default.

peterhoeg commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -1,57 +1,99 @@
-  name = "syncthing-${version}";
+let
+  common =
+    args@{ name, target, installPhase }:
```
1. You don't need ```args@```
2. Please join the 2 lines to become:
```
common = { name, target, installPhase }:
```

andrew-d committed:
Review feedback: simplify and move things around, add syncthing-cli package

andrew-d committed:
Make the build more pure by explicitly setting the build username/host

andrew-d commented:
@peterhoeg - Okay, I believe that I've addressed all of your feedback - let me know if I missed anything.  During the process, I also noticed a bit of an impurity in the build - the build script sets [linker flags][lf] based on the current username and host name, so I've added a commit to override those to both be `nix`.

[lf]: https://github.com/syncthing/syncthing/blob/v0.14.44/build.go#L742-L743

peterhoeg commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -1,57 +1,115 @@
-    ln -s $(pwd) src/github.com/syncthing/syncthing
-    export GOPATH=$(pwd)
+    installPhase = ''
+      mkdir -p $out/bin
```
You don't need ```mkdir``` here. ```install -D``` will take care of it.

peterhoeg commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -1,57 +1,115 @@
+    target = "stcli";
+
+    installPhase = ''
+      mkdir -p $out/bin
```
Or here

peterhoeg commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -1,57 +1,115 @@
+    target = "stdiscosrv";
+
+    installPhase = ''
+      mkdir -p $out/bin
```
Or here

peterhoeg commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -1,57 +1,115 @@
-    find $out/bin -type f -exec remove-references-to -t ${go} '{}' '+'
-  '';
+    installPhase = ''
+      mkdir -p $out/bin
```
Or here

peterhoeg commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -1,57 +1,115 @@
-  version = "0.14.44";
-  name = "syncthing-${version}";
+let
+  common = { stname, target, installPhase, patches ? null}:
```
```patches``` is a list so it should be ```[] ``` instead of ```null```.

peterhoeg reviewed (changes requested):
We've officially gone into nitpick territory now - thanks for your patience!

andrew-d committed:
Review feedback: remove spurious 'mkdir's and fix default argument

andrew-d commented:
@peterhoeg - Changes made! 😀

peterhoeg commented:
> current username and host name, so I've added a commit to override those to both be nix.

Nicely spotted!

I had a look at everything in its entirety and noticed that for every derivation, you essentially call what is:

```
install -Dm755 ${target} $out/bin/${target}
```

So that can be standardized in the ```common``` function and the 2 derivations that do more can instead pass in a ```postInstall``` for the other bits. That would rip out a few lines and reduce duplication.

andrew-d committed:
Review feedback: reduce duplication with a postInstall hook

andrew-d commented:
Ah, yeah, that's a good idea - just pushed a commit that does that 👍

andrew-d commented:
Awesome!  Thank you very much for the reviews 😀
